File: src/f2a/stats/missing.py
"""결측치 분석 모듈."""

from __future__ import annotations

import pandas as pd

from f2a.core.schema import DataSchema


class MissingStats:
    """결측치 패턴을 분석합니다.

    Args:
        df: 분석 대상 DataFrame.
        schema: 데이터 스키마.
    """

    def __init__(self, df: pd.DataFrame, schema: DataSchema) -> None:
        self._df = df
        self._schema = schema

    def column_summary(self) -> pd.DataFrame:
        """컬럼별 결측치 요약을 반환합니다.

        Returns:
            각 컬럼의 결측 수, 비율, 타입 정보를 포함한 DataFrame.
        """
        rows: list[dict] = []
        for col_info in self._schema.columns:
            rows.append(
                {
                    "column": col_info.name,
                    "missing_count": col_info.n_missing,
                    "missing_ratio": col_info.missing_ratio,
                    "missing_%": round(col_info.missing_ratio * 100, 2),
                    "dtype": col_info.dtype,
                }
            )

        result = pd.DataFrame(rows).set_index("column")
        return result.sort_values("missing_count", ascending=False)

    def row_missing_distribution(self) -> pd.DataFrame:
        """행 단위 결측 수 분포를 반환합니다.

        Returns:
            행별 결측 수의 빈도표.
        """
        row_missing = self._df.isna().sum(axis=1)
        dist = row_missing.value_counts().sort_index()
        return pd.DataFrame(
            {
                "missing_per_row": dist.index,
                "row_count": dist.values,
                "row_%": (dist.values / len(self._df) * 100).round(2),
            }
        )

    def missing_matrix(self) -> pd.DataFrame:
        """결측치 매트릭스 (boolean)를 반환합니다.

        시각화에 사용되는 결측 여부 행렬입니다.

        Returns:
            결측이면 True인 boolean DataFrame.
        """
        return self._df.isna()

    def total_missing_ratio(self) -> float:
        """전체 결측 비율을 반환합니다."""
        total_cells = self._df.shape[0] * self._df.shape[1]
        if total_cells == 0:
            return 0.0
        return round(float(self._df.isna().sum().sum() / total_cells), 4)

File: src/stats/missing.rs
use polars::prelude::*;
use serde::{Deserialize, Serialize};

use crate::core::schema::DataSchema;

// ─── Result types ───────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MissingColumnInfo {
    pub column: String,
    pub missing_count: usize,
    pub missing_ratio: f64,
    pub dtype: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MissingResult {
    /// Per-column missing information.
    pub columns: Vec<MissingColumnInfo>,
    /// Number of rows with at least one missing value.
    pub rows_with_missing: usize,
    /// Overall missing ratio (total missing / total cells).
    pub overall_missing_ratio: f64,
    /// Distribution of missing count per row: (n_missing, count_of_rows).
    pub row_missing_distribution: Vec<(usize, usize)>,
    /// Boolean missing matrix (columns × is_missing) for pattern analysis.
    /// Stored as column_name → Vec<bool> for serialization.
    pub missing_matrix: Vec<(String, Vec<bool>)>,
}

// ─── Implementation ─────────────────────────────────────────────────

pub struct MissingStats<'a> {
    df: &'a DataFrame,
    schema: &'a DataSchema,
}

impl<'a> MissingStats<'a> {
    pub fn new(df: &'a DataFrame, schema: &'a DataSchema) -> Self {
        Self { df, schema }
    }

    pub fn compute(&self) -> MissingResult {
        let n_rows = self.df.height();
        let n_cols = self.df.width();
        let total_cells = n_rows * n_cols;

        // ── Per-column info ─────────────────────────────────────
        let columns: Vec<MissingColumnInfo> = self
            .df
            .get_columns()
            .iter()
            .map(|col| {
                let missing_count = col.null_count();
                let missing_ratio = if n_rows > 0 {
                    missing_count as f64 / n_rows as f64
                } else {
                    0.0
                };
                MissingColumnInfo {
                    column: col.name().to_string(),
                    missing_count,
                    missing_ratio,
                    dtype: format!("{:?}", col.dtype()),
                }
            })
            .collect();

        let total_missing: usize = columns.iter().map(|c| c.missing_count).sum();
        let overall_missing_ratio = if total_cells > 0 {
            total_missing as f64 / total_cells as f64
        } else {
            0.0
        };

        // ── Per-row missing count ───────────────────────────────
        let mut row_missing_counts = vec![0usize; n_rows];
        for col in self.df.get_columns() {
            let is_null = col.is_null();
            for (i, val) in is_null.into_iter().enumerate() {
                if val.unwrap_or(false) {
                    row_missing_counts[i] += 1;
                }
            }
        }

        let rows_with_missing = row_missing_counts.iter().filter(|&&c| c > 0).count();

        // Distribution: how many rows have 0 missing, 1 missing, etc.
        let max_missing = *row_missing_counts.iter().max().unwrap_or(&0);
        let mut distribution = vec![0usize; max_missing + 1];
        for &c in &row_missing_counts {
            distribution[c] += 1;
        }
        let row_missing_distribution: Vec<(usize, usize)> = distribution
            .into_iter()
            .enumerate()
            .filter(|(_, count)| *count > 0)
            .collect();

        // ── Missing matrix ──────────────────────────────────────
        // Only generate for columns that have any missing (to save memory)
        let missing_matrix: Vec<(String, Vec<bool>)> = self
            .df
            .get_columns()
            .iter()
            .filter(|col| col.null_count() > 0)
            .map(|col| {
                let mask: Vec<bool> = col
                    .is_null()
                    .into_iter()
                    .map(|v| v.unwrap_or(false))
                    .collect();
                (col.name().to_string(), mask)
            })
            .collect();

        MissingResult {
            columns,
            rows_with_missing,
            overall_missing_ratio,
            row_missing_distribution,
            missing_matrix,
        }
    }
}
